### PR TITLE
cdc: cherry pick #5144 into release-5.1

### LINKED ISSUE #1386: Owner OOM when upstream TiDB has too many tables
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.
- Prepare a TiDB cluster with 40k+ tables, create 8 changefeeds, each of which only replicates a few tables (< 10).

2. What did you expect to see?
- That replication goes smoothly, and that memory consumption is not affected by the number of unreplicated tables.

3. What did you see instead?
- Owner runs out of memory.

4. Versions of the cluster

    - TiCDC version (execute `cdc version`):

        ```
        5.0.0-rc
        ```

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

### What problem does this PR solve?

Issue Number: close #1386 .

### What is changed and how it works?

Cherry pick #5144 into release-5.1

#### Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

### Release note 

```release-note
None
```

### CODE REVIEW

File: cmd/util.go [hunk 1]
```diff
@@ -256,25 +256,22 @@ func verifyTables(credential *security.Credential, cfg *config.ReplicaConfig, st
...
 	}
 
-	for tID, tableName := range snap.CloneTables() {
-		tableInfo, exist := snap.TableByID(tID)
-		if !exist {
-			return nil, nil, errors.NotFoundf("table %d", tID)
-		}
-		if filter.ShouldIgnoreTable(tableName.Schema, tableName.Table) {
-			continue
+    snap.IterTables(true, func(tableInfo *model.TableInfo) {
```
Comment (hicqu) [inline cmd/util.go:264]: Please help me to verify parameter `includeIneligible`.

File: cdc/owner/schema.go [hunk 2]
```diff
@@ -67,21 +67,21 @@ func (s *schemaWrap4Owner) AllPhysicalTables() []model.TableID {
 	if s.allPhysicalTablesCache != nil {
 		return s.allPhysicalTablesCache
 	}
-	tables := s.schemaSnapshot.Tables()
-	s.allPhysicalTablesCache = make([]model.TableID, 0, len(tables))
-	for _, tblInfo := range tables {
+	// NOTE: it's better to pre-allocate the vector. However in the current implementation
+	// we can't know how many valid tables in the snapshot.
+	s.allPhysicalTablesCache = make([]model.TableID, 0)
+	s.schemaSnapshot.IterTables(true, func(tblInfo *model.TableInfo) {
```
Comment (hicqu) [inline cdc/owner/schema.go:73]: Please help me to verify parameter includeIneligible.

**Review by hi-rustin — APPROVED** [review]
Thanks!

**Review by amyangfei — APPROVED** [review]

**Pull request merged**